### PR TITLE
Apply deepseek cuda rope

### PR DESCRIPTION
**Pull request by ispobock**
## Motivation
The torch.compile issue described in #4959 arises from overriding the `forward` method of `CustomOp`.

When torch.compile is applied, the `_forward_method` of `CustomOp` is replaced with `forward_native` (see [here](https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/model_executor/cuda_graph_runner.py#L59)). The `forward` method uses `_forward_method` for dispatching. However, if the forward method is overridden in `CustomOp`, the overridden method is still dispatched to the `forward_cuda` logic during compilation, leading to unexpected behavior.

Co-debugging with @fzyzcjy .

## Analysis: No worries about errors in CUDA rope (by @fzyzcjy)

One may wonder whether there is real error in CUDA implementation of rope, and we just hide / workaround it by disabling the CUDA implementation and use torch native implementation when doing torch.compile. Thus here is a brief analysis.

The error comes from the `triton_poi_fused_3` line below. As we can see, in_ptr0, i.e. buf2, is of shape 24x1x64. However, in that kernel, we are doing `tl.load` for 0...13824 indices, which is out of bound. Therefore, this seems like a bug in torch 2.5.1's compile and is unrelated to the CUDA rope kernel implementations.

<details>

```python
triton_poi_fused_3 = async_compile.triton('triton_poi_fused_3', '''
import triton
import triton.language as tl
from triton.compiler.compiler import AttrsDescriptor

from torch._inductor.runtime import triton_helpers, triton_heuristics
from torch._inductor.runtime.triton_helpers import libdevice, math as tl_math
from torch._inductor.runtime.hints import AutotuneHint, ReductionHint, TileHint, instance_descriptor, DeviceProperties

@triton_heuristics.pointwise(
    size_hints=[16384], 
    filename=__file__,
    triton_meta={'signature': {0: '*bf16', 1: '*bf16', 2: 'i32'}, 'device': DeviceProperties(type='cuda', index=7, cc=90, major=9, regs_per_multiprocessor=65536, max_threads_per_multi_processor=2048, multi_processor_count=132), 'constants': {}, 'configs': [AttrsDescriptor(divisible_by_16=(0, 1, 2), equal_to_1=())]},
    inductor_meta={'autotune_hints': set(), 'kernel_name': 'triton_poi_fused_3', 'mutated_arg_names': [], 'no_x_dim': False, 'num_load': 1, 'num_reduction': 0, 'backend_hash': '4CCCA65D160E209BCA90DF8D7D98B2EF1474A5946F9D1F5951D0EE51BFFF8A87', 'are_deterministic_algorithms_enabled': False, 'assert_indirect_indexing': True, 'autotune_local_cache': True, 'autotune_pointwise': True, 'autotune_remote_cache': None, 'force_disable_caches': False, 'dynamic_scale_rblock': True, 'max_autotune': True, 'max_autotune_pointwise': False, 'min_split_scan_rblock': 256, 'spill_threshold': 16, 'store_cubin': False, 'coordinate_descent_tuning': True, 'coordinate_descent_search_radius': 1, 'coordinate_descent_check_all_directions': False},
    min_elem_per_thread=0
)
@triton.jit
def triton_poi_fused_3(in_ptr0, out_ptr0, xnumel, XBLOCK : tl.constexpr):
    xnumel = 13824
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = xindex < xnumel
    x0 = xindex
    tmp0 = tl.load(in_ptr0 + (x0), xmask).to(tl.float32)
    tl.store(out_ptr0 + (x0), tmp0, xmask)
''', device_str='cuda')

def call(args):
...
        buf0 = empty_strided_cuda((24, 1), (1, 24), torch.float32)
        buf1 = empty_strided_cuda((24, 512), (512, 1), torch.bfloat16)
        # Topologically Sorted Source Nodes: [contiguous, x, pow_1, variance, add, rsqrt, x_1, mul_1, x_2], Original ATen: [aten.clone, aten._to_copy, aten.pow, aten.mean, aten.add, aten.rsqrt, aten.mul]
        stream7 = get_raw_stream(7)
        triton_per_fused__to_copy_add_clone_mean_mul_pow_rsqrt_0.run(arg0_1, arg1_1, buf0, buf1, 24, 512, grid=grid(24), stream=stream7)
        del arg1_1
        buf2 = empty_strided_cuda((24, 1, 64), (64, 64, 1), torch.bfloat16)
        # Topologically Sorted Source Nodes: [apply_rope_pos_ids_cos_sin_cache_default], Original ATen: [aten.view]
        triton_poi_fused_view_1.run(buf1, arg0_1, buf2, 1536, grid=grid(1536), stream=stream7)
        buf3 = empty_strided_cuda((589824, ), (1, ), torch.bfloat16)
        # Topologically Sorted Source Nodes: [], Original ATen: []
        triton_poi_fused_2.run(arg3_1, buf3, 589824, grid=grid(589824), stream=stream7)
        buf4 = empty_strided_cuda((13824, ), (1, ), torch.bfloat16)
        # Topologically Sorted Source Nodes: [], Original ATen: []
        triton_poi_fused_3.run(buf2, buf4, 13824, grid=grid(13824), stream=stream7)
```

</details>

### Accuracy
```bash
python3 -m sglang.launch_server --model /dev/shm/DeepSeek-V3 --tp 8 --trust-remote-code --attention-backend fa3 --port 33333 --disable-radix
python3 benchmark/gsm8k/bench_sglang.py --num-questions 1400 --parallel 1400 --port 33333 --num-shots 8

Accuracy: 0.946
Invalid: 0.000
Latency: 381.832 s
Output throughput: 364.705 token/s
```

#### Kernel
before:
before_cuda_rope
after:
after_cuda_rope
35μm -> 1.7 μm

#### E2E
```bash
SGL_ENABLE_JIT_DEEPGEMM=0 python3 -m sglang.launch_server --model /dev/shm/DeepSeek-V3 --tp 8 --trust-remote-code --attention-backend fa3 --port 33333 --disable-radix
python3 -m sglang.bench_one_batch_server --model None --base-url http://localhost:33333 --batch-size 1 --input-len 512 --output-len 512
```
```bash
# This PR
batch size: 1
latency: 12.65 s
output throughput: 40.46 token/s
(input + output) throughput: 80.92 token/s

# main
batch size: 1
latency: 13.54 s
output throughput: 37.81 token/s
(input + output) throughput: 75.63 token/s
```
7% output throughput improvement for batch size 1.

### CODE REVIEW

**Comment by fzyzcjy** [comment]
Analysis appended

**Review by zhyncs — APPROVED** [review]

**Comment by zhyncs** [comment]
ref https://github.com/sgl-project/sglang/pull/4959 https://github.com/sgl-project/sglang/pull/4764 https://github.com/sgl-project/sglang/pull/4813

File: python/sglang/srt/layers/rotary_embedding.py [hunk 1]
```diff
@@ -645,7 +645,10 @@ def _compute_cos_sin_cache(self) -> torch.Tensor:
         cache = torch.cat((cos, sin), dim=-1)
         return cache
 
-    def forward(
+    def forward_hip(self, *args, **kwargs):
+        return self.forward_native(*args, **kwargs)
```
Comment (zhyncs) [inline python/sglang/srt/layers/rotary_embedding.py:649]:
```suggestion
        return self.forward_cuda(*args, **kwargs)
```

File: .github/workflows/pr-test-amd.yml [hunk 2]
```diff
@@ -95,6 +95,7 @@ jobs:
           docker exec ci_sglang pip uninstall sgl-kernel -y || true
           docker exec -w /sglang-checkout/sgl-kernel ci_sglang bash -c "rm -f pyproject.toml && mv pyproject_rocm.toml pyproject.toml && python3 setup_rocm.py install"
           docker exec ci_sglang pip install -e "python[dev_hip]"
+          docker exec ci_sglang pip install py-spy || true
```
Comment (zhyncs) [inline .github/workflows/pr-test-amd.yml]:
```suggestion
```

File: python/sglang/srt/layers/rotary_embedding.py [hunk 3]
```diff
@@ -645,7 +645,10 @@ def _compute_cos_sin_cache(self) -> torch.Tensor:
         cache = torch.cat((cos, sin), dim=-1)
         return cache
 
-    def forward(
+    def forward_hip(self, *args, **kwargs):
+        return self.forward_cuda(*args, **kwargs)
```
Comment (zhyncs) [inline python/sglang/srt/layers/rotary_embedding.py]:
```suggestion
        return self.forward_native(*args, **kwargs)
```

**Comment by HaiShaw** [comment]
@fzyzcjy `Therefore, this seems like a bug in torch 2.5.1's compile and is unrelated to the CUDA rope kernel implementations.` - we do for torch 2.6 now, any update?

**Comment by zhyncs** [comment]
@fzyzcjy @ispobock @HaiShaw 
```python3
    def forward(self, *args, **kwargs):
        if torch._dynamo.is_compiling:
            return self.forward_native(*args, **kwargs)
        if _is_cuda_available:
            return self.forward_cuda(*args, **kwargs)
        else:
            return self.forward_native(*args, **kwargs)
```
I've fixed the issue with this.

**Comment by zhyncs** [comment]
Thanks @slin1237 for the help.

**Comment by fzyzcjy** [comment]
@HaiShaw I have not tested 2.6 whether it errors, but feel free to run it and see whether it triggers error. Anyway, with @ispobock's change, the error will not be triggerred even on 2.5.1 because cuda branch is not executed when compiling.

**Comment by Fridge003** [comment]
@ispobock @zhyncs 
After this PR, I tried to profile with 
```bash
python3 -m sglang.launch_server --model /dev/shm/DeepSeek-V3 --tp 8 --trust-remote-code --attention-backend fa3
```

But the result shows the Rope module is still using `forward_native`:
截屏2025-04-14 22 40 43

Do you have any clues?

**Comment by fzyzcjy** [comment]
@Fridge003 wondering whether there is a tiny typo, changed in https://github.com/sgl-project/sglang/pull/5406 (directly edited on github, not tested locally yet)
